File: backend/GameApp.Domain/ValueObjects/Scenes/TradeScene.cs
using GameApp.Domain.Enumerates;
using GameApp.Domain.ValueObjects.Items;

namespace GameApp.Domain.ValueObjects.Scenes;

// Class for scene where you trade with a merchant
public class TradeScene : Scene
{
    // Character offer
    public List<Item> CharacterItemsOffer { get; private set; } = new();
    public int CharacterMoneyOffer { get; private set; }


    // Merchant offer
    public List<Item> MerchantItemsOffer { get; private set; } = new();
    public int MerchantMoneyOffer { get; private set; }


    // Constructor
    public TradeScene(
        SceneName name,
        SceneDescription description,
        Biomes biome,
        List<Item> characterItemsOffer,
        int characterMoneyOffer,
        List<Item> merchantItemsOffer,
        int merchantMoneyOffer
    ) : base(name, description, biome)
    {
        CharacterItemsOffer = characterItemsOffer;
        CharacterMoneyOffer = characterMoneyOffer;
        MerchantItemsOffer = merchantItemsOffer;
        MerchantMoneyOffer = merchantMoneyOffer;
    }

    // Getters
    public List<Item> GetCharacterItemsOffer() => CharacterItemsOffer;
    public int GetCharacterMoneyOffer() => CharacterMoneyOffer;
    public List<Item> GetMerchantItemsOffer() => MerchantItemsOffer;
    public int GetMerchantMoneyOffer() => MerchantMoneyOffer;

    // Setters 
    public TradeScene SetSceneName(SceneName newName) =>
        new TradeScene(newName, GetDescription(), GetBiome(), CharacterItemsOffer, CharacterMoneyOffer, MerchantItemsOffer, MerchantMoneyOffer);

    public TradeScene SetSceneDescription(SceneDescription newDescription) =>
        new TradeScene(GetName(), newDescription, GetBiome(), CharacterItemsOffer, CharacterMoneyOffer, MerchantItemsOffer, MerchantMoneyOffer);

    public TradeScene SetBiome(Biomes newBiome) =>
        new TradeScene(GetName(), GetDescription(), newBiome, CharacterItemsOffer, CharacterMoneyOffer, MerchantItemsOffer, MerchantMoneyOffer);


    public TradeScene SetCharacterItemsOffer(List<Item> newItemsOffer) =>
        new TradeScene(GetName(), GetDescription(), GetBiome(), newItemsOffer, CharacterMoneyOffer, MerchantItemsOffer, MerchantMoneyOffer);

    public TradeScene SetCharacterMoneyOffer(int newMoneyOffer) =>
        new TradeScene(GetName(), GetDescription(), GetBiome(), CharacterItemsOffer, newMoneyOffer, MerchantItemsOffer, MerchantMoneyOffer);

    public TradeScene SetMerchantItemsOffer(List<Item> newItemsOffer) =>
        new TradeScene(GetName(), GetDescription(), GetBiome(), CharacterItemsOffer, CharacterMoneyOffer, newItemsOffer, MerchantMoneyOffer);

    public TradeScene SetMerchantMoneyOffer(int newMoneyOffer) =>
        new TradeScene(GetName(), GetDescription(), GetBiome(), CharacterItemsOffer, CharacterMoneyOffer, MerchantItemsOffer, newMoneyOffer);


    // To string
    public override string ToString()
    {
        string characterItems = string.Join(", ", CharacterItemsOffer.Select(i => i.ToString()));
        string merchantItems = string.Join(", ", MerchantItemsOffer.Select(i => i.ToString()));

        return $"TradeScene: {GetName()} - {GetBiome()}\n" +
               $"Character Offer: Money={CharacterMoneyOffer}, Items=[{characterItems}]\n" +
               $"Merchant Offer: Money={MerchantMoneyOffer}, Items=[{merchantItems}]";
    }
}

File: frontend/gameapp-frontend/src/domain/value-objects/scenes/trade-scene.js
import { Item } from "../items/item";
import { Scene } from "./scene";

export class TradeScene extends Scene {
    constructor(name, description, biome, characterItemsOffer = [], characterMoneyOffer = 0, merchantItemsOffer = [], merchantMoneyOffer = 0) {
        super(name, description, biome);

        this.#validateItemList(characterItemsOffer, "characterItemsOffer");
        this.#validateItemList(merchantItemsOffer, "merchantItemsOffer");

        this._characterItemsOffer = [...characterItemsOffer];
        this._characterMoneyOffer = characterMoneyOffer;
        this._merchantItemsOffer = [...merchantItemsOffer];
        this._merchantMoneyOffer = merchantMoneyOffer;
    }

    #validateItemList(list, paramName) {
        if (!Array.isArray(list)) throw new TypeError(`${paramName} must be an array`);
        for (const item of list) {
            if (!(item instanceof Item)) {
                throw new TypeError(`All elements of ${paramName} must be instances of Item or its subclasses`);
            }
        }
    }

    // getter
    get characterItemsOffer() { return [...this._characterItemsOffer]; }
    get characterMoneyOffer() { return this._characterMoneyOffer; }
    get merchantItemsOffer() { return [...this._merchantItemsOffer]; }
    get merchantMoneyOffer() { return this._merchantMoneyOffer; }

    // setter
    setSceneName(newName) { return new TradeScene(newName, this._description, this._biome, this._characterItemsOffer, this._characterMoneyOffer, this._merchantItemsOffer, this._merchantMoneyOffer); }
    setSceneDescription(newDescription) { return new TradeScene(this._name, newDescription, this._biome, this._characterItemsOffer, this._characterMoneyOffer, this._merchantItemsOffer, this._merchantMoneyOffer); }
    setBiome(newBiome) { return new TradeScene(this._name, this._description, newBiome, this._characterItemsOffer, this._characterMoneyOffer, this._merchantItemsOffer, this._merchantMoneyOffer); }

    setCharacterItemsOffer(newItems) {
        return new TradeScene(this._name, this._description, this._biome, newItems, this._characterMoneyOffer, this._merchantItemsOffer, this._merchantMoneyOffer);
    }
    setCharacterMoneyOffer(newMoney) { return new TradeScene(this._name, this._description, this._biome, this._characterItemsOffer, newMoney, this._merchantItemsOffer, this._merchantMoneyOffer); }
    setMerchantItemsOffer(newItems) {
        return new TradeScene(this._name, this._description, this._biome, this._characterItemsOffer, this._characterMoneyOffer, newItems, this._merchantMoneyOffer);
    }
    setMerchantMoneyOffer(newMoney) { return new TradeScene(this._name, this._description, this._biome, this._characterItemsOffer, this._characterMoneyOffer, this._merchantItemsOffer, newMoney); }

    toString() {
        const characterItems = this._characterItemsOffer.map(i => i.toString()).join(", ");
        const merchantItems = this._merchantItemsOffer.map(i => i.toString()).join(", ");
        return `TradeScene: ${this._name} - ${this._biome}\nCharacter Offer: Money=${this._characterMoneyOffer}, Items=[${characterItems}]\nMerchant Offer: Money=${this._merchantMoneyOffer}, Items=[${merchantItems}]`;
    }
}
